service/storage_proxy.cc

```diff
@@ -4861,26 +4863,26 @@ void storage_proxy::init_messaging_service() {
         });
     };
 
-    auto receive_mutation_handler = [] (const rpc::client_info& cinfo, rpc::opt_time_point t, frozen_mutation in, std::vector<gms::inet_address> forward,
+    auto receive_mutation_handler = [] (write_hint mutation_is_hint, const rpc::client_info& cinfo, rpc::opt_time_point t, frozen_mutation in, std::vector<gms::inet_address> forward,
             gms::inet_address reply_to, unsigned shard, storage_proxy::response_id_type response_id, rpc::optional<std::optional<tracing::trace_info>> trace_info) {
         tracing::trace_state_ptr trace_state_ptr;
         auto src_addr = netw::messaging_service::get_source(cinfo);
 
         utils::UUID schema_version = in.schema_version();
         return handle_write(src_addr, t, schema_version, std::move(in), std::move(forward), reply_to, shard, response_id,
                 trace_info ? *trace_info : std::nullopt,
-                /* apply_fn */ [] (shared_ptr<storage_proxy>& p, tracing::trace_state_ptr tr_state, schema_ptr s, const frozen_mutation& m,
+                /* apply_fn */ [mutation_is_hint] (shared_ptr<storage_proxy>& p, tracing::trace_state_ptr tr_state, schema_ptr s, const frozen_mutation& m,
                         clock_type::time_point timeout) {
-                    return p->mutate_locally(std::move(s), m, std::move(tr_state), db::commitlog::force_sync::no, timeout);
+                    return p->mutate_locally(std::move(s), m, std::move(tr_state), db::commitlog::force_sync::no, timeout, mutation_is_hint);
                 },
                 /* forward_fn */ [] (shared_ptr<storage_proxy>& p, netw::messaging_service::msg_addr addr, clock_type::time_point timeout, const frozen_mutation& m,
                         gms::inet_address reply_to, unsigned shard, response_id_type response_id,
                         std::optional<tracing::trace_info> trace_info) {
                     return p->_messaging.send_mutation(addr, timeout, m, {}, reply_to, shard, response_id, std::move(trace_info));
                 });
     };
-    ms.register_mutation(receive_mutation_handler);
-    ms.register_hint_mutation(receive_mutation_handler);
+    ms.register_mutation(std::bind_front<>(receive_mutation_handler, write_hint::no));
+    ms.register_hint_mutation(std::bind_front<>(receive_mutation_handler, write_hint::yes));
 
```

Comment:
Maybe just pass a pointer to the service group, instead of yes/no and then make a choice later? Saves a choice.

More complicated is to make smp_service_group scheduling_group aware.

Reply:
I need to check it but the mutate_locally functions are public while the smp service groups are private in storage proxy,
the question is if it justifies the split.
As an alternative to what I did here I can:
1. make the groups public ( as mutate locally is used outside of storage_proxy)
2. make a private mutate_locally functions that are being called with default smp groups by a public counterparts
3. Add an array of smp_groups that is indexed by an enum, this enum will be passed to the mutate_locally function

I prefer option 3 as it will externalize the concept without externalizing storage_proxy's internals.

Reply:
If we have a scheduling group aware smp_service_group it will also work for workload prioritization.

Reply:
I think option 2 is reasonable, it's nicer to have a different function rather than booleans that slightly affect behavior. Option 3 would be good if we didn't have the extra complication of workload prioritization.

Reply:
I don't have any specific problem with option 2, but I don't understand what does workload prioritization have got to do 
with it. It is not a good solution to create a semaphore per scheduling group because it will either increase the amount of 
cross shard operations that are active or will split the existing cross shard operations quota in a way that will not fully 
utilize them. I have just published a proposal [here](https://docs.google.com/document/d/1KJd-VRD_ZinlVbLDA6_qYfFjshcLAN9EaksoX0vB8xo/edit?usp=sharing) for a priority aware semaphore which should solve this problem for workload prioritization and for other cases like hints vs normal writes.

Reply:
Workload prioritization suffers from the same issue. If service level X stuffs the smp_service_group with requests, which are then handled slowly because X has few shares, then service level Y will block on the queue even though it has plenty of shares. 

If we have a priority aware sempahore then we can use it here, too, no?

Reply:
Ahh sure, I probably misunderstood you.
Will start working on an RFC for priority aware semaphore ASAP.

Reply:
I'll merge this, and we can replace with the super duper semaphore when we have it.